graphql_compiler/compiler/ir_lowering_match.py

```diff
@@ -584,6 +676,282 @@ def prune_output_blocks_in_compound_match_query(compound_match_query):
+        # Apply all filters for a location to the first occurence of that location
+        if match_step.as_block.location in location_to_filters:
+            where_block = Filter(
+                _filter_list_to_conjunction_expression(
+                    location_to_filters[match_step.as_block.location]
+                )
+            )
+            # Delete the location entry. No further filters needed for this location.
+            # If the same location is found in another call to this function,
+            # no filters will be added.
+            del location_to_filters[match_step.as_block.location]
+        else:
+            where_block = None
+
+        new_match_step = MatchStep(
+            root_block=match_step.root_block,
+            coerce_type_block=match_step.coerce_type_block,
+            where_block=where_block,
+            as_block=match_step.as_block
+        )
+        new_match_traversal.append(new_match_step)
+    return new_match_traversal
+
+
+def collect_filters_to_first_location_instance(compound_match_query):
+    """Collect all filters for a particular location to the first instance of the location."""
+    # Adding edge field non-exsistence filters in `_prune_traverse_using_omitted_locations`
+    # may result in filters being applied to locations after their first occurence.
+    # OrientDB does not resolve this behavior correctly.
+    # Therefore, for each MatchQuery, we collect all the filters for each location in a list.
+    # For each location, we make a conjunction of the filter list (`_predicate_list_to_where_block`)
+    # and apply the new filter to only the first instance of that location.
+    # All other instances will have no filters (None).
+    new_match_queries = []
+    # Each MatchQuery has a different set of locations, and associated Filters.
+    # Hence, each of them is processed independently.
+    for match_query in compound_match_query.match_queries:
+        # Construct mapping from location -> list of filter predicates applied at that location
+        location_to_predicates = _construct_location_to_filter_list(match_query)
+
+        new_match_traversals = []
+        for match_traversal in match_query.match_traversals:
+            new_match_traversal = _collect_filters_to_first_location_in_match_traversal(
+                match_traversal, location_to_predicates)
+            new_match_traversals.append(new_match_traversal)
+
+        new_match_queries.append(
+            MatchQuery(
+                match_traversals=new_match_traversals,
+                folds=match_query.folds,
+                output_block=match_query.output_block
+            )
+        )
+
+    return CompoundMatchQuery(match_queries=new_match_queries)
+
+
+def _update_context_field_binary_composition(expression, present_locations):
+    """Lower BinaryCompositions involving non-existent ContextFields to True.
+
+    Args:
+        expression: BinaryComposition with at least one ContextField operand
```

Comment:
If you require that the `BinaryComposition` be such that it has `at least one ContextField operand`, you should verify that precondition in the function. Right now, the function is just as happy with any kind of BinaryComposition, and will always return a value.